web/src/channel.js

```diff
@@ -90,13 +126,39 @@ function call(args) {
                 );
             }
         }
-        orig_error(xhr, error_type, xhn);
+        let callback_span;
+        if (span !== undefined) {
+            callback_span = span.startChild({
+                op: "function",
+                description: "error callback",
+            });
+            finish_span = () => {
+                callback_span.finish();
+                span.finish();
+            };
+        }
+        try {
+            orig_error(xhr, error_type, xhn);
+            if (callback_span !== undefined) {
+                callback_span.setStatus(SpanStatus.Ok);
+            }
+        } catch (error) {
+            if (callback_span !== undefined) {
+                callback_span.setStatus(SpanStatus.UnknownError);
+            }
+            throw error;
+        } finally {
+            finish_span();
+        }
```

Comment:
This kind of wrapping won’t be possible when we migrate from jQuery AJAX to the promise-based `fetch` API (after resolving #4650 in some way). Is it important?

Reply:
Specifically wrapping the callback?  We need to defer closing the top-most span (which they call a "transaction") until all of the parts we wanted included in it have finished -- which includes the HTTP call returning, so we know how long that took.

Actually wrapping the callback itself isn't all that important, but we do need to be able to hook into "after the request completes" somehow.

Reply:
The success and error callbacks are not logically part of the HTTP call. The only reason it’s mechanically part of the call at present is that callbacks were once the only way to write asynchronous APIs. In the future it will look more like

```js
try {
    const response = await fetch(…);
    // handle success
} catch (error) {
    // handle error
}
```

So what I think I’m saying is, until we get there, `channel` should finish any spans at the beginning of its own success and error callbacks, before invoking the corresponding user callbacks.

Reply:
Ah, OK, I see.  Yes, capturing the time spent in the callers' continuations only makes sense before we have the await.  It seemed useful now, since we _can_ automatically capture that time -- but we won't later, so to be able to do apples-to-apples comparisons later, we should push that responsibility onto the caller to document it themselves.

This also means there isn't much data duration which is captured in this span that isn't captured in the automatically-instrumented HTTP call (e.g. [Sentry link](https://zulip.sentry.io/discover/zulip-frontend:44c9eb2e763345e1af9c91141ff36ccb/?environment=czo&field=title&field=event.type&field=project&field=user.display&field=timestamp&field=replayId&name=All+Events&project=4504556882821120&query=user.id%3A699&sort=-timestamp&statsPeriod=48h&yAxis=count%28%29), for those with access):
<img width="1068" alt="Screenshot 2023-05-08 at 10 17 06 AM" src="https://user-images.githubusercontent.com/28347/236848166-7f676860-89b0-4b99-b3d5-556f00231fb4.png">

Reply:
Given that Sentry already does it's own spans for the HTTP requests themselves, I would argue that measuring the `success` call handler times is the most useful thing for us to add -- I certainly care about instrumenting many cases of "Rendering after an API call takes a long time", because those often turn out to be important web UI performance bugs.

Reply:
I totally agree that the time spent in the success callback is useful.  But fundamentally there _won't_ be a success callback later, so we should push the responsibility for adding that span to the caller.